torch/_inductor/simple_fsdp.py

```diff
@@ -0,0 +1,154 @@
+    Any,
+    Counter,
+    DefaultDict,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+)
+from enum import Enum
+from . import config, ir, comms, scheduler
+from collections import OrderedDict
+
+import torch
+
+from .dependencies import WeakDep
+from .utils import is_collective, is_wait, tuple_sorted
+
+import torch.distributed._functional_collectives
+_c10d_functional = torch.ops._c10d_functional
+
+class NodeType(Enum):
+    ALL_GATHER = 0
+    WAIT = 1
+    COMP = 2
+    REDUCE_SCATTER = 3
+    COVERT_ELEMENT = 4
+
+def reorder_all_gather(
+    snodes: List["scheduler.BaseSchedulerNode"],
+    all_gather_before_last_wait: Optional[bool] = True,
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_before_last_wait set to True: all_gather_i is reordered before wait_i-1
+    2. all_gather_before_last_wait set to False: all_gather_i is reordered after wait_i-1
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
+    all_gather_list: List["scheduler.BaseSchedulerNode"] = []
+    node_type_list: List[int] = []
+
+    inverse_users, node_users = comms.compute_node_users(snodes)
+
+    for node in reversed(snodes):
+        node_type_list.append(get_node_type(node))
+
+    for idx, node in enumerate(reversed(snodes)):
+        node_type = node_type_list[idx]
+        if node_type in [NodeType.REDUCE_SCATTER, NodeType.COMP]:
+            # we do not reorder reduce scatter and computation node
+            if node not in reorder_list and node not in all_gather_list:
+                reorder_list.append(node)
+        elif node_type == NodeType.ALL_GATHER:
+            # gather i-th all gather node and its dependencies
+            all_gather_list.append(node)
+            inverse_user = list(inverse_users[node])
+            if len(inverse_user) > 0:
+                all_gather_list.extend(inverse_user)
+        elif node_type == NodeType.WAIT:
+            if not all_gather_before_last_wait and node_type_list[idx+1]==NodeType.ALL_GATHER and len(all_gather_list) > 0:
+                # move i-th all gather node and its dependencies after (i-1)-th wait node (bc this is a reverse list)
+                reorder_list.extend(all_gather_list)
+                all_gather_list = []
+            # add wait node
+            reorder_list.append(node)
+            if all_gather_before_last_wait and node_type_list[idx+1]==NodeType.ALL_GATHER and len(all_gather_list) > 0: 
+                # move i-th all gather node and its dependencies before (i-1)-th wait node (bc this is a reverse list)
+                reorder_list.extend(all_gather_list)
+                all_gather_list = []
+        else:
+            raise ValueError("node type not supported")
+    if len(all_gather_list) > 0:
+        reorder_list.extend(all_gather_list)
+    reorder_list.reverse()
+    return reorder_list
+
+
+
```

Comment:
Let's use only 2 empty lines between functions. In fact, you can install `lintrunner` to help you do linting.